File: JasSwift/Sources/Algorithms/MagicWand.swift
// Magic Wand match predicate.
//
// Pure function: given a seed element, a candidate element, and the
// nine `state.magic_wand_*` configuration values, decide whether
// the candidate is "similar" to the seed under the enabled
// criteria.
//
// See `transcripts/MAGIC_WAND_TOOL.md` §Predicate for the rules.
// Cross-language parity is mechanical — the OCaml / Rust / Python
// ports of this module use the same logic.

import Foundation

/// The five-criterion configuration mirrors `state.magic_wand_*`.
/// Each criterion has an enabled flag (true = participate in the
/// predicate) and, where applicable, a tolerance.
public struct MagicWandConfig: Equatable {
    /// Fill Color criterion enabled.
    public var fillColor: Bool
    /// Maximum Euclidean RGB distance on the 0–255 scale.
    public var fillTolerance: Double

    /// Stroke Color criterion enabled.
    public var strokeColor: Bool
    /// Maximum Euclidean RGB distance on the 0–255 scale.
    public var strokeTolerance: Double

    /// Stroke Weight criterion enabled.
    public var strokeWeight: Bool
    /// Maximum |Δ width| in pt.
    public var strokeWeightTolerance: Double

    /// Opacity criterion enabled.
    public var opacity: Bool
    /// Maximum |Δ opacity| × 100 in percentage points.
    public var opacityTolerance: Double

    /// Blending Mode criterion enabled. Exact-match only.
    public var blendingMode: Bool

    public init(
        fillColor: Bool = true,
        fillTolerance: Double = 32,
        strokeColor: Bool = true,
        strokeTolerance: Double = 32,
        strokeWeight: Bool = true,
        strokeWeightTolerance: Double = 5.0,
        opacity: Bool = true,
        opacityTolerance: Double = 5,
        blendingMode: Bool = false
    ) {
        self.fillColor = fillColor
        self.fillTolerance = fillTolerance
        self.strokeColor = strokeColor
        self.strokeTolerance = strokeTolerance
        self.strokeWeight = strokeWeight
        self.strokeWeightTolerance = strokeWeightTolerance
        self.opacity = opacity
        self.opacityTolerance = opacityTolerance
        self.blendingMode = blendingMode
    }
}

/// Decide whether `candidate` is similar to `seed` under the
/// enabled criteria. AND across all enabled criteria. When all
/// criteria are disabled this returns false (the wand is a
/// no-op; the click handler treats this as "select only the seed
/// itself" — but that's the *caller's* responsibility).
public func magicWandMatch(
    seed: Element,
    candidate: Element,
    config: MagicWandConfig
) -> Bool {
    let anyEnabled = config.fillColor || config.strokeColor
        || config.strokeWeight || config.opacity || config.blendingMode
    if !anyEnabled { return false }

    if config.fillColor
        && !mwFillColorMatches(seed.fill, candidate.fill,
                               tolerance: config.fillTolerance) {
        return false
    }
    if config.strokeColor
        && !mwStrokeColorMatches(seed.stroke, candidate.stroke,
                                 tolerance: config.strokeTolerance) {
        return false
    }
    if config.strokeWeight
        && !mwStrokeWeightMatches(seed.stroke, candidate.stroke,
                                  tolerance: config.strokeWeightTolerance) {
        return false
    }
    if config.opacity
        && !mwOpacityMatches(mwElementOpacity(seed),
                             mwElementOpacity(candidate),
                             tolerance: config.opacityTolerance) {
        return false
    }
    if config.blendingMode
        && seed.blendMode != candidate.blendMode {
        return false
    }
    return true
}

/// Top-level opacity accessor for an Element. Each variant carries
/// its own opacity field; default is 1.0 across the model.
public func mwElementOpacity(_ e: Element) -> Double {
    switch e {
    case .line(let v): return v.opacity
    case .rect(let v): return v.opacity
    case .circle(let v): return v.opacity
    case .ellipse(let v): return v.opacity
    case .polyline(let v): return v.opacity
    case .polygon(let v): return v.opacity
    case .path(let v): return v.opacity
    case .text(let v): return v.opacity
    case .textPath(let v): return v.opacity
    case .group(let v): return v.opacity
    case .layer(let v): return v.opacity
    case .live(let v):
        switch v {
        case .compoundShape(let cs): return cs.opacity
        }
    }
}

// MARK: - Per-criterion helpers

private func mwFillColorMatches(_ seed: Fill?, _ cand: Fill?,
                                 tolerance: Double) -> Bool {
    switch (seed, cand) {
    case (nil, nil): return true
    case (.some(let s), .some(let c)):
        return mwRgbDistance(s.color.toRgba(), c.color.toRgba()) <= tolerance
    default: return false
    }
}

private func mwStrokeColorMatches(_ seed: Stroke?, _ cand: Stroke?,
                                   tolerance: Double) -> Bool {
    switch (seed, cand) {
    case (nil, nil): return true
    case (.some(let s), .some(let c)):
        return mwRgbDistance(s.color.toRgba(), c.color.toRgba()) <= tolerance
    default: return false
    }
}

private func mwStrokeWeightMatches(_ seed: Stroke?, _ cand: Stroke?,
                                    tolerance: Double) -> Bool {
    switch (seed, cand) {
    case (nil, nil): return true
    case (.some(let s), .some(let c)):
        return abs(s.width - c.width) <= tolerance
    default: return false
    }
}

private func mwOpacityMatches(_ seed: Double, _ cand: Double,
                               tolerance: Double) -> Bool {
    return abs(seed - cand) * 100.0 <= tolerance
}

/// Euclidean RGB distance on the 0-255 scale. Inputs are
/// `Color.toRgba()` outputs (R, G, B, A) in `[0.0, 1.0]`; we scale
/// R, G, B to `[0, 255]` and ignore alpha (Fill / Stroke carry
/// their own opacity field).
private func mwRgbDistance(_ a: (Double, Double, Double, Double),
                            _ b: (Double, Double, Double, Double)) -> Double {
    let dr = (a.0 - b.0) * 255.0
    let dg = (a.1 - b.1) * 255.0
    let db = (a.2 - b.2) * 255.0
    return (dr * dr + dg * dg + db * db).squareRoot()
}

File: jas_ocaml/lib/algorithms/magic_wand.ml
(* Magic Wand match predicate. See magic_wand.mli + MAGIC_WAND_TOOL.md
   §Predicate. *)

type config = {
  fill_color : bool;
  fill_tolerance : float;
  stroke_color : bool;
  stroke_tolerance : float;
  stroke_weight : bool;
  stroke_weight_tolerance : float;
  opacity : bool;
  opacity_tolerance : float;
  blending_mode : bool;
}

let default_config = {
  fill_color = true;
  fill_tolerance = 32.0;
  stroke_color = true;
  stroke_tolerance = 32.0;
  stroke_weight = true;
  stroke_weight_tolerance = 5.0;
  opacity = true;
  opacity_tolerance = 5.0;
  blending_mode = false;
}

(* Element accessors. Variants without fill / stroke / opacity report
   sensible defaults (no fill, no stroke, opacity 1.0) so the
   predicate can run on any element variant. *)
let fill_of (e : Element.element) : Element.fill option =
  match e with
  | Element.Rect { fill; _ }
  | Element.Circle { fill; _ }
  | Element.Ellipse { fill; _ }
  | Element.Polyline { fill; _ }
  | Element.Polygon { fill; _ }
  | Element.Path { fill; _ }
  | Element.Text { fill; _ }
  | Element.Text_path { fill; _ } -> fill
  | _ -> None

let stroke_of (e : Element.element) : Element.stroke option =
  match e with
  | Element.Line { stroke; _ }
  | Element.Rect { stroke; _ }
  | Element.Circle { stroke; _ }
  | Element.Ellipse { stroke; _ }
  | Element.Polyline { stroke; _ }
  | Element.Polygon { stroke; _ }
  | Element.Path { stroke; _ }
  | Element.Text { stroke; _ }
  | Element.Text_path { stroke; _ } -> stroke
  | _ -> None

let opacity_of (e : Element.element) : float =
  match e with
  | Element.Line { opacity; _ }
  | Element.Rect { opacity; _ }
  | Element.Circle { opacity; _ }
  | Element.Ellipse { opacity; _ }
  | Element.Polyline { opacity; _ }
  | Element.Polygon { opacity; _ }
  | Element.Path { opacity; _ }
  | Element.Text { opacity; _ }
  | Element.Text_path { opacity; _ }
  | Element.Group { opacity; _ }
  | Element.Layer { opacity; _ } -> opacity
  | _ -> 1.0

(* Euclidean RGB distance on the 0..255 scale. Inputs are
   [Element.color_to_rgba] outputs (R, G, B, A) in [0.0, 1.0]; we
   scale R, G, B to [0, 255] and ignore alpha — Fill / Stroke carry
   their own [opacity] field. *)
let rgb_distance (a : Element.color) (b : Element.color) : float =
  let (ar, ag, ab, _) = Element.color_to_rgba a in
  let (br, bg, bb, _) = Element.color_to_rgba b in
  let dr = (ar -. br) *. 255.0 in
  let dg = (ag -. bg) *. 255.0 in
  let db = (ab -. bb) *. 255.0 in
  sqrt (dr *. dr +. dg *. dg +. db *. db)

let fill_color_matches
    (seed : Element.fill option) (cand : Element.fill option)
    (tolerance : float) : bool =
  match seed, cand with
  | None, None -> true
  | Some s, Some c -> rgb_distance s.fill_color c.fill_color <= tolerance
  | _ -> false

let stroke_color_matches
    (seed : Element.stroke option) (cand : Element.stroke option)
    (tolerance : float) : bool =
  match seed, cand with
  | None, None -> true
  | Some s, Some c ->
    rgb_distance s.stroke_color c.stroke_color <= tolerance
  | _ -> false

let stroke_weight_matches
    (seed : Element.stroke option) (cand : Element.stroke option)
    (tolerance : float) : bool =
  match seed, cand with
  | None, None -> true
  | Some s, Some c -> Float.abs (s.stroke_width -. c.stroke_width) <= tolerance
  | _ -> false

let opacity_matches (seed : float) (cand : float) (tolerance : float) : bool =
  Float.abs (seed -. cand) *. 100.0 <= tolerance

let blending_mode_matches
    (seed : Element.blend_mode) (cand : Element.blend_mode) : bool =
  seed = cand

let magic_wand_match
    (seed : Element.element) (candidate : Element.element)
    (cfg : config) : bool =
  let any_enabled =
    cfg.fill_color || cfg.stroke_color || cfg.stroke_weight
    || cfg.opacity || cfg.blending_mode
  in
  if not any_enabled then false
  else if cfg.fill_color
       && not (fill_color_matches (fill_of seed) (fill_of candidate)
                 cfg.fill_tolerance)
  then false
  else if cfg.stroke_color
       && not (stroke_color_matches (stroke_of seed) (stroke_of candidate)
                 cfg.stroke_tolerance)
  then false
  else if cfg.stroke_weight
       && not (stroke_weight_matches (stroke_of seed) (stroke_of candidate)
                 cfg.stroke_weight_tolerance)
  then false
  else if cfg.opacity
       && not (opacity_matches (opacity_of seed) (opacity_of candidate)
                 cfg.opacity_tolerance)
  then false
  else if cfg.blending_mode
       && not (blending_mode_matches
                 (Element.get_blend_mode seed)
                 (Element.get_blend_mode candidate))
  then false
  else true
